internal/resolution/variablesources/installed_package_test.go

```diff
@@ -4,392 +4,439 @@ import (
+				Image:   "registry.io/repo/test-package@v2.2.0",
+				Properties: []property.Property{
+					{Type: property.TypePackage, Value: json.RawMessage(`{"packageName": "test-package", "version": "2.2.0"}`)},
+				},
+			},
+			InChannels: []*catalogmetadata.Channel{&testPackageChannel},
+		},
+		// We need a bundle from different package to ensure that
+		// we filter out bundles certain bundle image
+		"some-other-package.v2.3.0": {
+			Bundle: declcfg.Bundle{
+				Name:    "some-other-package.v2.3.0",
+				Package: "some-other-package",
+				Image:   "registry.io/repo/some-other-package@v2.3.0",
+				Properties: []property.Property{
+					{Type: property.TypePackage, Value: json.RawMessage(`{"packageName": "some-other-package", "version": "2.3.0"}`)},
+				},
+			},
+			InChannels: []*catalogmetadata.Channel{&someOtherPackageChannel},
+		},
+	}
+	allBundles := make([]*catalogmetadata.Bundle, 0, len(bundleSet))
+	for _, bundle := range bundleSet {
+		allBundles = append(allBundles, bundle)
+	}
 
-	t.Run("UpgradeConstraintPolicy is set to Ignore", func(t *testing.T) {
-		const bundleImage = "registry.io/repo/test-package@v2.0.0"
-		fakeOperator := fakeOperator("test-operator", "test-package", operatorsv1alpha1.UpgradeConstraintPolicyIgnore)
-		installedPackages, err := variablesources.MakeInstalledPackageVariables(
-			allBundles,
-			[]operatorsv1alpha1.Operator{fakeOperator},
-			[]rukpakv1alpha1.BundleDeployment{
-				fakeBundleDeployment("test-package-bd", bundleImage, &fakeOperator),
+	for _, tt := range []struct {
+		name                    string
+		upgradeConstraintPolicy operatorsv1alpha1.UpgradeConstraintPolicy
+		installedBundle         *catalogmetadata.Bundle
+		expectedResult          []*olmvariables.InstalledPackageVariable
+		expectedError           string
+	}{
+		{
+			name:                    "respect replaces directive from catalog",
+			upgradeConstraintPolicy: operatorsv1alpha1.UpgradeConstraintPolicyEnforce,
+			installedBundle:         bundleSet["test-package.v2.0.0"],
+			expectedResult: []*olmvariables.InstalledPackageVariable{
+				olmvariables.NewInstalledPackageVariable(testPackageName, []*catalogmetadata.Bundle{
+					// Must only have two bundle:
+					// - the one which replaces the current version
+					// - the current version (to allow to stay on the current version)
+					bundleSet["test-package.v2.1.0"],
+					bundleSet["test-package.v2.0.0"],
+				}),
+			},
+		},
+		{
+			name:                    "UpgradeConstraintPolicy is set to Ignore",
+			upgradeConstraintPolicy: operatorsv1alpha1.UpgradeConstraintPolicyIgnore,
+			installedBundle:         bundleSet["test-package.v2.0.0"],
+			expectedResult:          []*olmvariables.InstalledPackageVariable{},
+		},
+		{
+			name:                    "no BundleDeployment for an Operator",
+			upgradeConstraintPolicy: operatorsv1alpha1.UpgradeConstraintPolicyEnforce,
+			expectedResult:          []*olmvariables.InstalledPackageVariable{},
+		},
```

Comment:
Could you explain these two tests?

Reply:
@awgreene sure. 

1. `UpgradeConstraintPolicy is set to Ignore` - When `UpgradeConstraintPolicy` is set to `Ignore` we basically saying that "there are no constraints on how to ugprade". This allows upgrades/downgrades to any version.
2. `no BundleDeployment for an Operator` - means that there is an `Operator` CR, but the requested package has not yet been installed and there is no `BundleDeployment`. In this case - we do not apply upgrade constraints as well (most likely we are reconciling the installation, not an upgrade/downgrade).

Please don't hesitate to to suggest alternative names for test cases if you think that we can communicate better the porpuse of these test cases.